crds/gen.go

```diff
@@ -0,0 +1,73 @@
+package main
+
+import (
+	"fmt"
+	"io"
+)
+
+const pulumiName = "pulumi"
+const kubernetesName = "k8s"
+
+// Maps the OpenAPI types to TypeScript types
+var types = map[string]string{
+	"integer": "number",
+	"number":  "number",
+	"string":  "string",
+	"boolean": "boolean",
+}
+
+// Wraps a type around the Pulumi input type. For example, inputType("number")
+// returns "pulumi.Input<number>" if const pulumiName = "pulumi".
+func inputType(T string) string {
+	return fmt.Sprintf("%s.Input<%s>", pulumiName, T)
+}
+
+// Returns the modulePath relative to the variable for the imported Pulumi
+// module. For example, if const pulumiName = "pulumi", then
+// fromPulumi("CustomResourceOptions") = "pulumi.CustomResourceOptions"
+func fromPulumi(modulePath string) string {
+	return fmt.Sprintf("%s.%s", pulumiName, modulePath)
+}
+
+// Returns the modulePath relative to the variable for the imported
+// pulumi-kubernetes module. For example, if const kubernetesName = "k8s", then
+// fromKubernetes("apiextensions.CustomResource") =
+// "k8s.apiextensions.CustomResource"
+func fromKubernetes(modulePath string) string {
+	return fmt.Sprintf("%s.%s", kubernetesName, modulePath)
+}
+
+func (r ResourceDefinition) GenerateNodeJS(w io.Writer) {
```

Comment:
This is a good start for learning, but you'll eventually want to generate a Pulumi schema internally so that you can use our codegen packages.

[Here's the schema](https://github.com/pulumi/pulumi-random/blob/master/provider/cmd/pulumi-resource-random/schema.json) for a simple provider to give you an idea of what I'm talking about.

We have types defined already for this schema, so you can translate from the `map[string]interface{}` you decode from the YAML and convert directly to the schema types.

Schema types are defined [here](https://github.com/pulumi/pulumi/blob/master/pkg/codegen/schema/schema.go)
Example usage is [here](https://github.com/pulumi/pulumi-kubernetes/blob/master/provider/pkg/gen/schema.go)